# HG changeset patch
# User John W. Eaton <jwe@octave.org>
# Date 1455651592 18000
#      Tue Feb 16 14:39:52 2016 -0500
# Node ID 987c1a79d33fa8ea99df82e9e797c7852afd6de2
# Parent  7e67c7f82fc1724cd5f9b61b3808f8efc4b557c2
new set_internal_variable variant

* variables.h, variables.cc (set_internal_variable): Set a string
variable limited to a list of choices.

diff --git a/libinterp/corefcn/variables.cc b/libinterp/corefcn/variables.cc
--- a/libinterp/corefcn/variables.cc
+++ b/libinterp/corefcn/variables.cc
@@ -964,16 +964,59 @@ set_internal_variable (int& var, const o
         }
       if (i == nchoices)
         error ("%s: value not allowed (\"%s\")", nm, sval.c_str ());
     }
 
   return retval;
 }
 
+octave_value
+set_internal_variable (std::string& var, const octave_value_list& args,
+                       int nargout, const char *nm, const char **choices)
+{
+  octave_value retval;
+  int nchoices = 0;
+  while (choices[nchoices] != 0)
+    nchoices++;
+
+  int nargin = args.length ();
+
+  if (nargout > 0 || nargin == 0)
+    retval = var;
+
+  if (wants_local_change (args, nargin))
+    {
+      if (! try_local_protect (var))
+        warning ("\"local\" has no effect outside a function");
+    }
+
+  if (nargin > 1)
+    print_usage ();
+
+  if (nargin == 1)
+    {
+      std::string sval = args(0).xstring_value ("%s: first argument must be a string", nm);
+
+      int i = 0;
+      for (; i < nchoices; i++)
+        {
+          if (sval == choices[i])
+            {
+              var = sval;
+              break;
+            }
+        }
+      if (i == nchoices)
+        error ("%s: value not allowed (\"%s\")", nm, sval.c_str ());
+    }
+
+  return retval;
+}
+
 struct
 whos_parameter
 {
   char command;
   char modifier;
   int parameter_length;
   int first_parameter_length;
   int balance;
diff --git a/libinterp/corefcn/variables.h b/libinterp/corefcn/variables.h
--- a/libinterp/corefcn/variables.h
+++ b/libinterp/corefcn/variables.h
@@ -111,16 +111,20 @@ set_internal_variable (double& var, cons
                        double minval = -octave_Inf,
                        double maxval = octave_Inf);
 
 extern OCTINTERP_API octave_value
 set_internal_variable (std::string& var, const octave_value_list& args,
                        int nargout, const char *nm, bool empty_ok = true);
 
 extern OCTINTERP_API octave_value
+set_internal_variable (std::string& var, const octave_value_list& args,
+                       int nargout, const char *nm, const char **choices);
+
+extern OCTINTERP_API octave_value
 set_internal_variable (int& var, const octave_value_list& args,
                        int nargout, const char *nm, const char **choices);
 
 #define SET_INTERNAL_VARIABLE(NM) \
   set_internal_variable (V ## NM, args, nargout, #NM)
 
 #define SET_NONEMPTY_INTERNAL_STRING_VARIABLE(NM) \
   set_internal_variable (V ## NM, args, nargout, #NM, false)
